### ISSUE TITLE
MyApp in production serves a jQuery and other debug-only resources

### ISSUE THREAD

**Issue by guest20** [kraih/mojo#1094]
```
$ mojo version
CORE
  Perl        (v5.18.2, darwin)    ## Mac OS X - 10.10.5 (14F2109)
  Mojolicious (7.30, Doughnut)
 - - 8< - -
```

### Steps to reproduce the behaviour
```
$ ./script/newly_generated_app get /mojo/jquery/jquery.js
$ MOJO_MODE=production hypnotoad  newly_generated_app/script/newly_generated_app
[Tue May 16 22:41:50 2017] [info] Listening at "http://*:8080"
Server available at http://127.0.0.1:8080
$  curl -q  http://localhost:8080/mojo/jquery/jquery.js  2>/dev/null |head -n1
/*! jQuery v3.1.1 | (c) jQuery Foundation | jquery.org/license */
```

### Expected behaviour
I would expect jQuery, and other `debug.html.ep`-only resources to 404 when the application is in production mode.

### Actual behaviour
You get jQuery 3.1.1:
`/*! jQuery v3.1.1 | (c) jQuery Foundation | jquery.org/license */`

Looking at the template one could imagine replacing the jQuery usage with plain-old-javascript as it's only a couple of hide/shows, but it might make more sense to arrange that content of that directory be served by a plugin/helper that's only loaded when the app is in development mode.

Thoughts?

**Comment by marcusramberg** [comment]
Does this cause any actual problems in production?

**Comment by guest20** [comment]
It's not a huge issue, it just seems a tad unsanitary to have debug artefacts publicly exposed in my prod environment. It's a bit of a lurking corner case.

I could very well just ask nginx to not proxy requests to /mojo to the app at all, but it seems like an unneeded layer of complication for every deployment to include (since I'd need a different config for nginx in prod/dev). 

For those who are into jQuery and are using it as part of their app, perhaps at a newer version there's now this goofy case where there are 2 versions of jquery on the site, which could be enough to allow bad guys some kind of foot hold (say an XSS that loads the older one to get back an old jquery bug).

Would y'all be open to a pull request that "fixes" this with a similar mode switch to the one logging uses?

**Comment by jhthorsen** [comment]
I must say I disagree that this is an issue.

👎 on doing anything special to prevent jQuery from being accessible in production.

**Comment by guest20** [comment]
How about doing something special to make it accessible in development?

**Comment by guest20** [comment]
I made a first pass at this by simply disabling the bundled resources via an attribute in: https://github.com/guest20/mojo/commit/a29386270b180846df4775b7243f25858141da3f

Clearly mojolicious.org (for example) can't switch this on because that particular site uses all the stuff in `resources`, but for the sites that don't need the mojo logo in pride-of-place this should be ok. This will also remove the favicon and break the default production error pages (neither bother me as I have already provided my own).

Is this patch too disruptive?

**Comment by jhthorsen** [comment]
I still don't understand the reason for this change.

**Comment by marcusramberg** [comment]
I'm :-1: on this as well. If you're bothered, I suggest blocking the assets
in the frontend as you suggested.

**Comment by guest20** [comment]
I don't understand the confusion. What value does serving these things bring to my site? Why is a simple "please don't" switch too disruptive?

**Comment by kraih** [comment]
This proposal would break http://mojolicious.org, which reuses some of the bundled assets.

**Comment by kraih** [comment]
Same for the documentation browser, which i assume some folks use in production too.

**Comment by jberger** [comment]
Ah the documentation browser is a good counterpoint. If that is using bundled assets (and it must be now that I think of it) then there really is no way to conditionally disable the bundled assets (say by mode) which is how I was thinking of doing it.

**Comment by guest20** [comment]
For the record, I don't want to block chunks of my url space, because my application serves search results from the 'not found' page, and hosting a search result for `mojo` at `/mojo` could indeed be a useful thing for a blog. In future I may even want `/mojo/` to refer to all articles with the tag "mojo".

@kraih's "mojolicious.org can't turn this on" is 100% on point, and I think that "just don't switch this on" applies just as well for that application as it does for the pod browser itself. 

This is the thinking behind the attribute's default - "yes, please serve these resources". Nobody will even notice the change unless they pass it a true value, at which point they can simply copy the files they'd like to keep from `@INC//Mojolicious/resources` into their own project's `/public` directory.

To address @jberger's point of the pod browser serving its own assets - I can enable/disable that by enabling/disabling the `::Plugin`, which is perfect. No plugin, none of its depends.

Sorry if I missed covering these points earlier.

**Comment by s1037989** [comment]
I imagine a CPAN plugin could be created that, when plugged in, would hook and disable those resources when the mode is production or not development.  So to turn off these resources would be as simple as:

```
$self->plugin('NoMojo');
```

Is the issue a matter of which is the better default?  Explicitly turn them on in production versus explicitly turn them off in production?  It does make sense for them to be on by default in production because even production produces default exception and not_found pages.  If you prefer to provide your own templates for this, simply plugin "NoMojo".

**Comment by guest20** [comment]
@s1037989, The "NoMojo" plugin would need to keep up with the content of `resources`, and would need to hook `before_static`, only to say "nah, don't dispatch this to static, re-dispatch this to the app, without accounting for static". To me that sounds like some pretty hard-core routing for the sake of avoiding a 2 line patch that turns those things off on request. 

I'd be happy to see how simple that plugin would be, but it sounds like quite a contortion.

**Comment by guest20** [comment]
@s1037989, generally removing `/mojo` by default makes sense to me, and seems like the kind of thing one would want once they have their own custom error templates. 

It turns out existing stuff uses things in that directory, and it doesn't make sense to break everyone else's stuff by default.

**Comment by kraih** [comment]
I don't know if this feature should be implemented or not, i don't feel very strongly about it either way. I'm just raising issues a final proposal needs to address, you can't break our website or apps that use the documentation browser. Everything else will be up to votes from the community.

**Comment by kraih** [comment]
It appears you also forgot about bundled templates.

**Comment by guest20** [comment]
I'm not at all bothered by bundled templates, they don't lurk in my url space outside my apps dispatch.

I did come across `mojo/perldoc` in my adventures, but that requires that I load `::PODRenderer` in order for it to be the target of dispatch - at least as far as I understand.

As a side note: It seems like that template could live in `*DATA` in `:: PODRenderer` without any loss of functionality.

**Comment by guest20** [comment]
@kraih besides `mojo/t` are there any other tests I should be running to find out if my changes break other things?

**Comment by niczero** [comment]
@guest20 Are you still putting together a proposal or can this be closed?

**Comment by guest20** [comment]
What is required for this "proposal"? As far as I see, it's simply:

"I'd like a switch so that I can turn off serving the bundled assets in my app".

Is there some level of detail I'm missing?

**Comment by kraih** [comment]
Btw. There is a pretty reasonable use case for disabling bundled static files. Right now you can't actually serve a `favicon.ico` from a route.

**Comment by guest20** [comment]
Any news on this?

Would something as simple as `has serve_bundled => sub { 1 };` be enough to address this?

**Comment by kraih** [comment]
@guest20 No that would not work. And i'm afraid there have been no new developments since this thread went idle.

**Comment by kraih** [comment]
To be more specific, a simple `serve_bundled` switch would break features that depend on those bundled assets, like our famous failraptor on the fallback 500 page. This needs a well thought out solution.

**Comment by guest20** [comment]
It seems quite clear that turning off bundled assets would break things that depend on them.  However my (possibly faulty) assumption here is that one who would switch off the bundled assets would be doing so because they want exactly this kind of control.

Do you think it would be sufficient to simply document the switch with "This feature completely disables bundled assets, which breaks the failraptor and some dev-only functionality. If you'd like to preserve them you can simply copy the bundled assets from %s into your apps asset directories"?

(The manual step could made easier with the addition of a `mojo generate bundled` or something similar.)

**Comment by kraih** [comment]
A handwavy `...which breaks the failraptor and some dev-only functionality` really doesn't cut it. We're talking about features that have their [own sections](http://mojolicious.org/perldoc/Mojolicious/Guides/Rendering#Rendering-exception-and-not_found-pages) in the guides and are used [all over the place](http://mojolicious.org/perldoc/Mojolicious/Guides/Cookbook#Subprocesses). Users might not even be aware [which of their plugins](http://grep.cpan.me/?q=reply%5C-%5C%3Enot_found) setting the flag might break.

**Comment by guest20** [comment]
What about the "copy the bundled bits into my app" part? Does that not give all the plugins that would be broken by not serving them from bundled the ability to find those templates/images/javascript again?

**Comment by kraih** [comment]
This has been resolved by #1113.
